From 4e57a3e1b6948774459970343d7f3f58d6cd4d50 Mon Sep 17 00:00:00 2001
From: kholoud mohamed <kholoudm@google.com>
Date: Mon, 29 Nov 2021 18:12:43 +0000
Subject: [PATCH] Fix Managed Provisioning crash on OTA

Fixed a crash that happened on OTA for devices that have other profile
types where ManagedProvisioning is not installed.

Bug: 207335612
Bug: 207573077
Fixes: 205276569
Bug: 195630721
Test: manual testing
Change-Id: I716b4aded965af7730860804646317eae7321d7b
Merged-In: I716b4aded965af7730860804646317eae7321d7b
(cherry picked from commit a01c0d7fa435694a76850b446c3efa03cd7a7677)
(cherry picked from commit e3027e5be5286b9d7c2e614759f7303126c80017)
(cherry picked from commit 2dba964ca701ee884ed3e9b02086d839fa8694e1)
Merged-In:I716b4aded965af7730860804646317eae7321d7b
---
 .../UpdateInteractAcrossProfilesAppOpTask.java | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/src/com/android/managedprovisioning/task/UpdateInteractAcrossProfilesAppOpTask.java b/src/com/android/managedprovisioning/task/UpdateInteractAcrossProfilesAppOpTask.java
index b3218fde..18993ab8 100644
--- a/src/com/android/managedprovisioning/task/UpdateInteractAcrossProfilesAppOpTask.java
+++ b/src/com/android/managedprovisioning/task/UpdateInteractAcrossProfilesAppOpTask.java
@@ -48,6 +48,7 @@ public class UpdateInteractAcrossProfilesAppOpTask extends AbstractProvisioningT
     private final DevicePolicyManager mDevicePolicyManager;
     private final AppOpsManager mAppOpsManager;
     private final PackageManager mPackageManager;
+    private final UserManager mUserManager;
 
     public UpdateInteractAcrossProfilesAppOpTask(Context context,
             ProvisioningParams provisioningParams,
@@ -59,6 +60,7 @@ public class UpdateInteractAcrossProfilesAppOpTask extends AbstractProvisioningT
         mDevicePolicyManager = context.getSystemService(DevicePolicyManager.class);
         mAppOpsManager = context.getSystemService(AppOpsManager.class);
         mPackageManager = context.getPackageManager();
+        mUserManager = context.getSystemService(UserManager.class);
     }
 
     @Override
@@ -133,10 +135,20 @@ public class UpdateInteractAcrossProfilesAppOpTask extends AbstractProvisioningT
 
     private Set<ApplicationInfo> getAllInstalledApps() {
         final Set<ApplicationInfo> apps = new HashSet<>();
-        List<UserHandle> profiles = mContext.getSystemService(UserManager.class).getAllProfiles();
+        List<UserHandle> profiles = mUserManager.getAllProfiles();
         for (UserHandle profile : profiles) {
-            apps.addAll(mContext.createContextAsUser(profile, /* flags= */ 0).getPackageManager()
-                    .getInstalledApplications(/* flags= */ 0));
+            if (profile.getIdentifier() != mContext.getUserId()
+                    && !mUserManager.isManagedProfile(profile.getIdentifier())) {
+                continue;
+            }
+            try {
+                apps.addAll(
+                        mContext.createPackageContextAsUser(
+                        /* packageName= */ "android", /* flags= */ 0, profile)
+                                .getPackageManager().getInstalledApplications(/* flags= */ 0));
+            } catch (PackageManager.NameNotFoundException ignored) {
+                // Should never happen.
+            }
         }
         return apps;
     }
-- 
2.17.1

